lib/api/model/model.dart

```diff
@@ -868,3 +880,84 @@ class DmMessage extends Message {
   @override
   Map<String, dynamic> toJson() => _$DmMessageToJson(this);
 }
+
+enum MessageEditState {
+  none,
+  edited,
+  moved;
+
+  /// When a topic is resolved, the clients agree on adding a ✔ prefix to the
+  /// topic string. When deciding whether a message has been moved or not, two
+  /// topics whose only difference is the ✔ prefix are considered the same.
+  static bool areSameTopic(String topic, String prevTopic) {
+    // TODO(#744) Extract this to its own home to fully support mark as resolve
+
+    // Code adapted from the shared code: web/shared/src/resolve_topic.ts
+
+    /**
+     * Pattern for an arbitrary resolved-topic prefix.
+     *
+     * These always begin with the canonical prefix, but can go on longer.
+     */
+    // The class has the same characters as RESOLVED_TOPIC_PREFIX.
+    // It's designed to remove a weird "✔ ✔✔ " prefix, if present.
+    final RegExp resolvedTopicPrefixRe = RegExp('^✔ [ ✔]*');
+
+    // Normalize both topics so the resolved-topic prefix do not interfere.
+    topic = topic.replaceFirst(resolvedTopicPrefixRe, '');
+    prevTopic = prevTopic.replaceFirst(resolvedTopicPrefixRe, '');
+
+    return topic != prevTopic;
+  }
+
+  static MessageEditState readFromMessage(Map<dynamic, dynamic> json, String key) {
+    return isMoved(json['edit_history'] as List<dynamic>?, json['last_edit_timestamp'] as int?);
+  }
+
+  static MessageEditState isMoved(List<dynamic>? editHistory, int? lastEditTimeStamp) {
```

Comment:
The intention was to make it easier for us to parse these information from `Message` instead of `Map<String, dynamic>` in the future. But yeah it can be inlined.